README.md
# Malicious Chrome Extension Scanner


## Code of Conduct
See [Contributor Covenant](CODE%20OF%20CONDUCT.md)


## Architecture
![Workflow Diagram](/images/Malicious%20Chrome%20Extension%20Scanner%20Workflow.png)
 
> We choose the Drive Route

**NOTE** Other ways to extract extensions, other than tenable, is present in [other ways to extract extension](/other_ways_to_extract_extensions)

## Description
To implement a allowlist policy for chrome extensions, it was necessary to get the chrome extensions used by the employees and go about evaluating the risk associated with the chrome extension. Hence, a combination of the extensions pulled from tenable and passing those extensions to **Crxcavator** (a Duo product that scans extensions) provided the starting step for analysis of the extensions.


## PreRequisites
To run the script, you would need to fill in the details in [config.ini](config.ini)

### Output: Drive Option 
As we used the **Drive** Option as Output:
1. A service account in GSuite needs to be created
2. A GCP Project with service account having the following roles:
    - https://www.googleapis.com/auth/drive.file 
    - https://www.googleapis.com/auth/drive.appdata 
    - https://www.googleapis.com/auth/drive.readonly
    - https://www.googleapis.com/auth/spreadsheets
3. Add the GSuite Service Account to the Team Drive so that it can create a Google Sheet for output
4. Activate the Drive API in the Google API Console
5. Activate the Sheet API in the Google API Console
More info abt setting up GSuite Auth in https://developers.google.com/identity/protocols/oauth2/service-account

**NOTE** GCP Service Account doesn't need domain wide delegation access 

### Setting up Tenable
1. The current script uses **tags** from Tenable and uses **Source** tag. If you haven't setup tags before in Tenable, see https://docs.tenable.com/tenableio/vulnerabilitymanagement/Content/Settings/Tags.htm
Once a new tag is setup, you would to wait from a few hours depending on the number of assets. 
2. You will also need to setup a scan that has Windows plugin `96533` & Mac plugin `133180` enabled and scans once at least every 7 days.

### Setting up Crxcavator
Create an account in https://crxcavator.io/ to generate a key you will need in config.ini


## Installing requirements
```shell script
pip install -r requirements.lock
```


## Output
The script will generate multiple sheets in GDrive under the mentioned team drive.
- _Risky_Chrome_Extensions_datetimestamp_
- _Paid_Chrome_Extensions_datetimestamp_
- _Chrome_Extensions_Removed_from_Store_datetimestamp_
- _Version_Unavailable_in_Crxcavator_datetimestamp_
- _Unscanned_in_Crxcavator_datetimestamp_  

**NOTE** The script was unable to fetch any results on the extensions mentioned in _Unscanned_in_Crxcavator_datetimestamp_.


## Options
##### Run ```python -B main.py --help``` to get a list of the options.
```buildoutcfg
Usage: main.py [OPTIONS]

Options:
  -d, --duration INTEGER  Duration in days, that you want to
                          pull the extensions from tenable
  --help                  Show this message and exit.
```


## Examples:  
1. To get all extensions since the last 7 days of tenable on workstations, use 
```shell script
python -B main.py -d 7
```

**NOTE** Assuming the workstations are scanned once in 7 days 


## Score Calculation
Score is calculated by taking the total of each section mentioned in https://crxcavator.io/docs#/risk_breakdown ignoring **External metadata** & **WebStore**. RetireJS is calculated in the total score even though its not mentioned in the column.


## Remove Extension
Some of the ways to remove extension can be via [GPO](https://www.tecklyfe.com/how-to-prevent-google-chrome-extensions-in-group-policy/), [powershell script](/ways_to_remove_extension/powershell.ps1), bash script, removing registry keys associated with chrome extensions etc.


## Credits
Malicious Chrome Extension Scanner is owned and maintained by [MAINTAINERS](MAINTAINERS.md)


## Contributing
See [CONTRIBUTING.md](CONTRIBUTING.md)


## License
Malicious Chrome Extension Scanner is licensed under 2-clause BSD License
See [LICENSE](LICENSE.md) for details.


## Release of new version
See [RELEASING.md](RELEASING.md)


## Need to report a vulnerability
See [SECURITY.md](SECURITY.md)


CHANGELOG.md
### v0.1.0
- This is the very first release!

drive.py
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry
import configparser
import json
import logging
import time
import jwt
from typing import Tuple


logger = logging.getLogger(__name__)

config = configparser.ConfigParser()
config.read('config.ini')

MAX_THREADS = int(config['drive_sheet']['threads'])
drive_api = config['drive_sheet']['drive_api']

drive_access_tokens = {}


# Generate session with max of 3 retries and interval of 1 second
def session_generator():
    session = requests.Session()
    retry = Retry(connect=3, backoff_factor=0.5)
    adapter = HTTPAdapter(max_retries=retry)
    session.mount('http://', adapter)
    session.mount('https://', adapter)
    return session


# Create OAuth token per requirement for each recipient
def generate_drive_api_access_token(recipient: str) -> Tuple[str, int]:
    access_token = None
    expiry = None
    jwt_header = {"alg": "RS256", "typ": "JWT"}
    iat = time.time()
    exp = iat + 3600
    jwt_claim_set = {
        'iss': config['drive_sheet']['gcp_project_service_account'],
        'scope': 'https://www.googleapis.com/auth/drive.file https://www.googleapis.com/auth/drive.appdata https://www.googleapis.com/auth/drive.readonly',
        'sub': recipient,
        'aud': 'https://www.googleapis.com/oauth2/v4/token',
        'iat': iat,
        'exp': exp
    }
    secret = bytes(config['drive_sheet']['key'].replace('\\n', '\n'), 'utf-8')
    signed_jwt = jwt.encode(jwt_claim_set, secret, headers=jwt_header, algorithm='RS256')
    headers = {"Content-Type": "application/json; charset=utf-8"}
    data = {'grant_type': 'urn:ietf:params:oauth:grant-type:jwt-bearer', 'assertion': signed_jwt.decode('utf-8').replace("'", '"')}
    url = 'https://www.googleapis.com/oauth2/v4/token'
    session = session_generator()
    resp = session.post(url, headers=headers, data=json.dumps(data))
    response = resp.json()
    if resp.ok:
        access_token = response['access_token']
        expiry = time.time() + response['expires_in']
    elif resp.status_code == 429:
        logger.error('Too many requests. Sleeping %s' % response['error_description'])
        time.sleep(1)
        access_token, expiry = generate_drive_api_access_token(recipient)
    elif 499 < resp.status_code < 600:
        logger.error('Server Error. Sleeping 1 second')
        time.sleep(1)
        access_token, expiry = generate_drive_api_access_token(recipient)
    else:
        logger.error('Failed to generate access token')
        logger.error("%d:%s" % (resp.status_code, resp.text))
    return access_token, expiry


def find_drive(drive_name: str, recipient: str) -> str:
    drive_id = None
    access_token = drive_access_tokens[recipient]['access_token']
    expiry = drive_access_tokens[recipient]['expiry']
    query_start_time = time.time()

    # Make the API call if token expiry time is greater than 1 minute
    if (expiry - query_start_time) > 60:
        headers = {'Content-Type': 'application/json', 'Authorization': 'Bearer %s' % access_token}
        url = drive_api.format("drives")
        session = session_generator()
        resp = session.get(url, headers=headers)
        if resp.ok:
            response = resp.json()
            for drive in response['drives']:
                if drive['kind'] == 'drive#drive' and drive['name'] == drive_name:
                    drive_id = drive['id']
        # Rate limiting
        elif resp.status_code == 429:
            logger.error('Too many requests. Sleeping %s' % resp.json()['error']['message'])
            time.sleep(1)
            drive_id = find_drive(drive_name, recipient)
        elif 499 < resp.status_code < 600:
            logger.error('Server Error. Sleeping 1 second')
            time.sleep(1)
            drive_id = find_drive(drive_name, recipient)
        else:
            logger.error('Failed to create folder')
            logger.error("%d:%s" % (resp.status_code, resp.text))
    # Create new access token to be used by the recipient
    else:
        access_token, expiry = generate_drive_api_access_token(recipient)
        if access_token is not None and expiry is not None:
            drive_access_tokens[recipient]['access_token'] = access_token
            drive_access_tokens[recipient]['expiry'] = expiry
            drive_id = find_drive(drive_name, recipient)
    return drive_id


def find_item(name: str, recipient: str, driveId: str, search_type: str, parent_id: str = None):
    _id = None
    access_token = drive_access_tokens[recipient]['access_token']
    expiry = drive_access_tokens[recipient]['expiry']
    query_start_time = time.time()

    # Make the API call if token expiry time is greater than 1 minute
    if (expiry - query_start_time) > 60:
        headers = {'Content-Type': 'application/json', 'Authorization': 'Bearer %s' % access_token}
        params = {
            'corpora': 'drive',
            'includeItemsFromAllDrives': True,
            'supportsAllDrives': True,
            'driveId': driveId,
            'fields': 'incompleteSearch,files(id,name,mimeType,parents)'
        }

        if parent_id is None:
            params['q'] = "name = '%s' and trashed = false" % name
        else:
            params['q'] = "name = '%s' and '%s' in parents and trashed = false" % (name, parent_id)

        if search_type == 'folder':
            params['q'] = "%s and mimeType = 'application/vnd.google-apps.folder'" % params['q']
        elif search_type == 'document':
            params['q'] = "%s and mimeType = 'application/vnd.google-apps.document'" % params['q']
        elif search_type == 'sheet':
            params['q'] = "%s and mimeType = 'application/vnd.google-apps.spreadsheet'" % params['q']
        elif search_type == 'presentation':
            params['q'] = "%s and mimeType = 'application/vnd.google-apps.presentation'" % params['q']

        url = drive_api.format("files")
        session = session_generator()
        resp = session.get(url, headers=headers, params=params)
        if resp.ok:
            response = resp.json()
            if response['incompleteSearch']:
                logger.warning('All files/folders are not returned.')
            if response['files']:
                if len(response['files']) == 1:
                    _id = response['files'][0]['id']
                else:
                    logger.warning('Too many folders found with the same name. Aborting search of folder and creating a new folder')
                    for file in response['files']:
                        print(file['name'], ':', file['id'])

        # Rate limiting
        elif resp.status_code == 429:
            logger.error('Too many requests. Sleeping %s' % resp.json()['error']['message'])
            time.sleep(1)
            _id = find_item(name, recipient, driveId, search_type, parent_id)
        elif 499 < resp.status_code < 600:
            logger.error('Server Error. Sleeping 1 second')
            time.sleep(1)
            _id = find_item(name, recipient, driveId, search_type, parent_id)
        else:
            logger.error('Failed to find folder %s' % name)
            logger.error("%d:%s" % (resp.status_code, resp.text))
    # Create new access token to be used by the recipient
    else:
        access_token, expiry = generate_drive_api_access_token(recipient)
        if access_token is not None and expiry is not None:
            drive_access_tokens[recipient]['access_token'] = access_token
            drive_access_tokens[recipient]['expiry'] = expiry
            _id = find_item(name, recipient, driveId, parent_id)
    return _id


def create_file(filename: str, recipient: str, drive_id: str, file_type: str, parent_folder_id: str = ""):
    if not parent_folder_id:
        parent_folder_id = drive_id
    _id = None
    access_token = drive_access_tokens[recipient]['access_token']
    expiry = drive_access_tokens[recipient]['expiry']
    query_start_time = time.time()

    if file_type == 'folder':
        mimeType = 'application/vnd.google-apps.folder'
    elif file_type == 'document':
        mimeType = 'application/vnd.google-apps.document'
    elif file_type == 'sheet':
        mimeType = 'application/vnd.google-apps.spreadsheet'
    elif file_type == 'presentation':
        mimeType = 'application/vnd.google-apps.presentation'
    else:
        mimeType = 'application/vnd.google-apps.folder'

    # Make the API call if token expiry time is greater than 1 minute
    if (expiry - query_start_time) > 60:
        headers = {'Content-Type': 'application/json', 'Authorization': 'Bearer %s' % access_token}
        file_metadata = {
            'name': filename,
            'mimeType': mimeType,
            'parents': [parent_folder_id],
            'driveId': drive_id
        }
        url = drive_api.format("files")
        session = session_generator()
        resp = session.post(url, headers=headers, json=file_metadata, params={'supportsAllDrives': True, 'fields': 'id'})
        if resp.ok:
            response = resp.json()
            if 'id' in response and response['id']:
                _id = response['id']
        # Rate limiting
        elif resp.status_code == 429:
            logger.error('Too many requests. Sleeping %s' % resp.json()['error']['message'])
            time.sleep(1)
            _id = create_file(filename, recipient, drive_id, file_type, parent_folder_id)
        elif 499 < resp.status_code < 600:
            logger.error('Server Error. Sleeping 1 second')
            time.sleep(1)
            _id = create_file(filename, recipient, drive_id, file_type, parent_folder_id)
        else:
            logger.error('Failed to create file: %s' % filename)
            logger.error("%d:%s" % (resp.status_code, resp.text))
    # Create new access token to be used by the recipient
    else:
        access_token, expiry = generate_drive_api_access_token(recipient)
        if access_token is not None and expiry is not None:
            drive_access_tokens[recipient]['access_token'] = access_token
            drive_access_tokens[recipient]['expiry'] = expiry
            _id = create_file(filename, recipient, drive_id, file_type, parent_folder_id)
    return _id


def copy_file(file_id: str, folder_id: str, recipient: str, drive_id: str):
    id = None
    access_token = drive_access_tokens[recipient]['access_token']
    expiry = drive_access_tokens[recipient]['expiry']
    query_start_time = time.time()

    # Make the API call if token expiry time is greater than 1 minute
    if (expiry - query_start_time) > 60:
        headers = {'Content-Type': 'application/json', 'Authorization': 'Bearer %s' % access_token}
        url = drive_api.format("files/%s/copy" % file_id)
        data = {
            'parents[]': [drive_id, folder_id]
        }
        session = session_generator()
        resp = session.post(url, headers=headers, json=data)
        response = resp.json()
        if resp.ok:
            id = response['id']
            # Rate limiting
        elif resp.status_code == 429:
            logger.error('Too many requests. Sleeping %s' % resp.json()['error']['message'])
            time.sleep(1)
            id = copy_file(file_id, folder_id, recipient, drive_id)
        elif 499 < resp.status_code < 600:
            logger.error('Server Error. Sleeping 1 second')
            time.sleep(1)
            id = copy_file(file_id, folder_id, recipient, drive_id)
        else:
            logger.error('Failed to copy file: %s' % file_id)
            logger.error("%d:%s" % (resp.status_code, resp.text))
    # Create new access token to be used by the recipient
    else:
        access_token, expiry = generate_drive_api_access_token(recipient)
        if access_token is not None and expiry is not None:
            drive_access_tokens[recipient]['access_token'] = access_token
            drive_access_tokens[recipient]['expiry'] = expiry
            id = copy_file(file_id, folder_id, recipient, drive_id)
    return id


def delete_file(file_id: str, recipient: str):
    access_token = drive_access_tokens[recipient]['access_token']
    expiry = drive_access_tokens[recipient]['expiry']
    query_start_time = time.time()

    # Make the API call if token expiry time is greater than 1 minute
    if (expiry - query_start_time) > 60:
        headers = {'Content-Type': 'application/json', 'Authorization': 'Bearer %s' % access_token}
        url = drive_api.format("files/%s" % file_id)
        session = session_generator()
        resp = session.delete(url, headers=headers)
        if resp.status_code == 429:
            logger.error('Too many requests. Sleeping %s' % resp.json()['error']['message'])
            time.sleep(1)
            delete_file(file_id, recipient)
        elif 499 < resp.status_code < 600:
            logger.error('Server Error. Sleeping 1 second')
            time.sleep(1)
            delete_file(file_id, recipient)
        elif not resp.ok:
            logger.error('Failed to delete file: %s' % file_id)
            logger.error("%d:%s" % (resp.status_code, resp.text))
    # Create new access token to be used by the recipient
    else:
        access_token, expiry = generate_drive_api_access_token(recipient)
        if access_token is not None and expiry is not None:
            drive_access_tokens[recipient]['access_token'] = access_token
            drive_access_tokens[recipient]['expiry'] = expiry
            delete_file(file_id, recipient)


ways_to_remove_extension/powershell.ps1
# Remove-ChromeExtension
# Remove Chrome extension by Extension ID and output results, will need to change code
# in order to be used with CrowdStrike Response API and our python scripts
#
######################

$TargetExtensionID = ""  # Enter Chrome Extension ID
$AllUserFolders = Get-ChildItem -Path "C:\Users"
ForEach ($UserFolder in $AllUserFolders){
    if (Test-Path -Path "$($UserFolder.FullName)\AppData\Local\Google\Chrome\User Data\Default\Extensions")
    {
        $AllExtensionsFolders = Get-ChildItem -Path "$($UserFolder.FullName)\AppData\Local\Google\Chrome\User Data\Default\Extensions"
        ForEach ($ExtensionFolder in $AllExtensionsFolders){
            $VersionFolders = Get-ChildItem -Path "$($ExtensionFolder.FullName)"
            Foreach ($VersionFolder in $VersionFolders) {
                $json = Get-Content -Raw -Path "$($VersionFolder.FullName)\manifest.json" | ConvertFrom-Json
                $ExtensionName = $json.name
                if( $ExtensionName -like "*MSG*" ) {
                    if( Test-Path -Path "$($VersionFolder.FullName)\_locales\en\messages.json" ) {
                        $json = Get-Content -Raw -Path "$($VersionFolder.FullName)\_locales\en\messages.json" | ConvertFrom-Json
                        $ExtensionName = $json.appName.message
                        if(!$ExtensionName) {
                            $ExtensionName = $json.extName.message
                        }
                        if(!$ExtensionName) {
                            $ExtensionName = $json.app_name.message
                        }
                    }

                    if( Test-Path -Path "$($VersionFolder.FullName)\_locales\en_US\messages.json" ) {
                        $json = Get-Content -Raw -Path "$($VersionFolder.FullName)\_locales\en_US\messages.json" | ConvertFrom-Json
                        $ExtensionName = $json.appName.message
                        if(!$ExtensionName) {
                            $ExtensionName = $json.extName.message
                        }
                        if(!$ExtensionName) {
                            $ExtensionName = $json.app_name.message
                        }
                    }
                }

                if( $TargetExtensionID -eq $ExtensionFolder) {
                    Remove-item -Path "$($UserFolder.FullName)\AppData\Local\Google\Chrome\User Data\Default\Extensions\($TargetExtensionID)" -Force -Recurse -ErrorAction SilentlyContinue
                    "Removed the following extension:"
                    @{Host=$env:computername;User=$UserFolder.Name;ExtensionID=$ExtensionFolder.Name;ExtensionName=$ExtensionName;Version=$VersionFolder.Name} | ConvertTo-Json -Compress
                }
                # Generating output as JSON, including hostname & username
            }
        }
    }
}

other_ways_to_extract_extensions/mac_script.sh
#!/bin/zsh
loggedInUser=$( stat -f "%Su" /dev/console )

path="/Users/${loggedInUser}/Library/Application Support/Google/Chrome/Default/Extensions"
if [[ $1 != "" ]]; then
  path=$1
  echo "Read extensions from \"$path\""
  echo
fi

JSONS=$( find "$path" -maxdepth 4 -name "manifest.json" )

while read JSON; do
    NAME=$( awk -F'"' '/name/{print $4}' "$JSON" )
    VERSION=$( awk -F'"' '/"version"/{print $4}' "$JSON" )
    DESCRIPTION=$( awk -F'"' '/"description"/{print $4}' "$JSON" )
    ID=$( echo "$JSON" | awk -F'/' '{print $(NF-2)}'  )
    EXT_PATH="$path/$ID"

    if [[ ! -z "$NAME" ]] && [[ ! "$NAME" =~ "_MSG_" ]]; then
        EXTS+=( "${NAME}\n" )
        echo "id: $ID"
        echo "name: $NAME"
        echo "version: $VERSION"
        echo "description: $DESCRIPTION"
        echo "path: $EXT_PATH"
        echo "url: $URL"
        echo
    fi
done < <(echo "$JSONS")

crxcavator.py
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry
import logging
import configparser
import time
import json
import concurrent.futures


logger = logging.getLogger(__name__)

config = configparser.ConfigParser()
config.read('config.ini')

crxcavator_api = config['crxcavator']['api']

MAX_THREADS = int(config['crxcavator']['threads'])  # Get max number of threads for multi-threading


class CrXcavator(object):
    def __init__(self, extension_id, version, name):
        self.id = extension_id
        self.version = version
        self.name = name
        self.risk_csp = None
        self.risk_external_javascript = None
        self.risk_external_calls = None
        self.risk_score = 0
        self.entry_points = None
        self.dangerous_functions = None
        self.chrome_link = "https://chrome.google.com/webstore/detail/{0}".format(extension_id)
        self.crxcavator_link = "https://crxcavator.io/report/{0}/{1}".format(extension_id, version)

    def print(self):
        print('ID: %s' % self.id)
        print('Version: %s' % self.version)
        print('Score: %d' % self.risk_score)
        print('Link: %s' % self.chrome_link)
        print('CrXcavator Link: %s' % self.crxcavator_link)
        if self.risk_csp is not None:
            print('CSP: \n%s' % json.dumps(self.risk_csp, indent=2))
        if self.risk_external_javascript is not None:
            print('External JavaScript: \n%s' % json.dumps(self.risk_external_javascript, indent=2))
        if self.risk_external_calls is not None:
            print('External Calls: \n%s' % json.dumps(self.risk_external_calls, indent=2))
        if self.dangerous_functions is not None:
            print('Dangerous Functions: \n%s' % json.dumps(self.dangerous_functions, indent=2))
        if self.entry_points is not None:
            print('Entry Points: \n%s' % json.dumps(self.entry_points, indent=2))
        print()


# Generate session with max of 3 retries and interval of 1 second
def session_generator():
    session = requests.Session()
    session.headers.update({'API-Key': config['crxcavator']['key'], 'Content-Type': 'application/json'})
    retry = Retry(connect=3, backoff_factor=0.5)
    adapter = HTTPAdapter(max_retries=retry)
    session.mount('http://', adapter)
    session.mount('https://', adapter)
    return session


# Parse risk data returned from report of crxcavator
def parse_risk_data(extension_id, version, data):
    riskobj = CrXcavator(extension_id, version, data['webstore']['name'])

    if 'csp' in data:
        riskobj.risk_csp = data['csp']

    if 'extjs' in data:
        riskobj.risk_external_javascript = data['extjs']

    if 'extcalls' in data:
        riskobj.risk_external_calls = data['extcalls']

    if 'entrypoints' in data:
        riskobj.entry_points = data['entrypoints']

    if 'dangerousfunctions' in data:
        riskobj.dangerous_functions = data['dangerousfunctions']

    if 'risk' in data:
        for each_item in data['risk']:
            if each_item == 'total' or each_item == 'webstore' or each_item == 'metadata':
                continue
            else:
                riskobj.risk_score = riskobj.risk_score + int(data['risk'][each_item]['total'])
    return riskobj


# Get risk data for a particular extension and their version
def get_extension_risk(extension_id, version):
    risk_obj = None
    session = session_generator()
    resp = session.get("%s/report/%s/%s" % (crxcavator_api, extension_id, version))
    if resp.ok:
        try:
            response = resp.json()
        except json.decoder.JSONDecodeError:
            logger.warning('JSON Decode Error. Retrying for extension %s version %s' % (extension_id, version))
            risk_obj = get_extension_risk(extension_id, version)
            return risk_obj

        if response is None:
            logger.info('Failed to fetch report on %s version %s' % (extension_id, version))
        else:
            if 'version' in response:
                if response['version'] is not None:
                    risk_obj = parse_risk_data(extension_id, response['version'], response['data'])
            else:
                print(json.dumps(response, indent=4))
    elif  600 > resp.status_code >= 500 or resp.status_code == 429:
        logger.warning("Exceed rate limit.")
        time.sleep(60)
        # TO DO:
        # Check header to see if spits out retry.
        # print(resp.header)
        risk_obj = get_extension_risk(extension_id, version)
    else:
        logger.error('ERROR %s: %s' % (resp.status_code, resp.text))
        logger.error('Unable to get risk data on extension %s of version %s' % (extension_id, version))
    return risk_obj


# Submit an extension to get it scanned by crxcavator. This would also be useful to classify the extensions to the
# below categories
def submit_extension(extension_id: str):
    submit_results = {}
    submit_results['id'] = extension_id
    submit_results['version'] = None
    submit_results['extension'] = False
    submit_results['not_free'] = False
    submit_results['run_again'] = False
    submit_results['removed_from_store'] = False

    data = {'extension_id': extension_id}
    session = session_generator()
    resp = session.post("%s/submit" % crxcavator_api, json=data)
    if resp.ok:
        try:
            response = resp.json()
        except json.decoder.JSONDecodeError:
            logger.warning('JSON Decode Error. Retrying for extension %s' % extension_id)
            submit_results = submit_extension(extension_id)
            return submit_results

        if 'error' not in response:
            if "no longer in Chrome" in response['message']:
                submit_results['removed_from_store'] = True
            else:
                submit_results['version'] = response['version']
                submit_results['extension'] = True
        else:
            if "not free" in response['error']:
                submit_results['not_free'] = True
            elif "Error retrieving extension from webstore" in response['error']:
                submit_results['run_again'] = True
            elif "Theme" in response['error']:
                submit_results['extension'] = False
            elif 'Error extension is too big' in response['error']:
                submit_results['version'] = ""
                submit_results['extension'] = True
            else:
                logger.error('Extension %s: %s' % (extension_id, response['error']))
    elif resp.status_code == 429:
        logger.warning("Exceed rate limit.")
        time.sleep(60)
        # TO DO:
        # Check header to see if spits out retry.
        # print(resp.header)
        submit_results = submit_extension(extension_id)
    elif 600 > resp.status_code >= 500:
        time.sleep(90)
        logger.error('Server not responsive for extension %s. Trying Again' % extension_id)
        submit_results['run_again'] = True
    else:
        logger.error('ERROR %s: %s' % (resp.status_code, resp.text))
    return submit_results


# Get risk data on multiple versions of the same chrome extension
def fetch_risk_details(extension_id, versions):
    riskobjs = []
    # Check if report exist for current version
    with concurrent.futures.ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
        fs = [executor.submit(get_extension_risk, extension_id, version) for version in versions]
        for future in concurrent.futures.as_completed(fs):
            riskobj = future.result()
            if riskobj is not None:
                riskobjs.append(riskobj)
    return riskobjs
